nexus/src/db/queries/external_ip.rs

```diff
@@ -1023,7 +1052,103 @@ mod tests {
     }
 
     #[tokio::test]
-    async fn test_insert_external_ip_for_service_is_idempoent() {
+    async fn test_explicit_external_ip_for_service_is_idempotent() {
+        let context = TestContext::new(
+            "test_explicit_external_ip_for_service_is_idempotent",
+        )
+        .await;
+
+        let ip_range = IpRange::try_from((
+            Ipv4Addr::new(10, 0, 0, 1),
+            Ipv4Addr::new(10, 0, 0, 4),
+        ))
+        .unwrap();
+        context.initialize_ip_pool("oxide-service-pool", ip_range).await;
+
+        // Allocate an IP address as we would for an external, rack-associated
+        // service.
+        let id = Uuid::new_v4();
+        let ip = context
+            .db_datastore
+            .allocate_explicit_service_ip(
+                &context.opctx,
+                id,
+                IpAddr::V4(Ipv4Addr::new(10, 0, 0, 3)),
+            )
+            .await
+            .expect("Failed to allocate service IP address");
+        assert_eq!(ip.kind, IpKind::Service);
+        assert_eq!(ip.ip.ip(), IpAddr::V4(Ipv4Addr::new(10, 0, 0, 3)));
+        assert_eq!(ip.first_port.0, 0);
+        assert_eq!(ip.last_port.0, u16::MAX);
+        assert!(ip.instance_id.is_none());
+
+        // Try allocating the same service IP again.
+        let ip_again = context
+            .db_datastore
+            .allocate_explicit_service_ip(
+                &context.opctx,
+                id,
+                IpAddr::V4(Ipv4Addr::new(10, 0, 0, 3)),
+            )
+            .await
+            .expect("Failed to allocate service IP address");
+        assert_eq!(ip.id, ip_again.id);
+        assert_eq!(ip.ip.ip(), ip_again.ip.ip());
+
+        // Try allocating the same service IP once more, but do it with a
+        // different input address.
+        let ip_again = context
+            .db_datastore
+            .allocate_explicit_service_ip(
+                &context.opctx,
+                id,
+                IpAddr::V4(Ipv4Addr::new(10, 0, 0, 2)),
+            )
+            .await
+            .expect("Failed to allocate service IP address");
+
+        // We return success, but with the address requested the first time.
+        assert_eq!(ip.id, ip_again.id);
+        assert_eq!(ip.ip.ip(), ip_again.ip.ip());
```

Comment:
This test was pretty surprising to me. The queries on line 1071 and 1087 ask for the address `10.0.0.3` twice, and it looks like we're testing that we get back the exact same record we've already inserted. That checks the query is idempotent, and looks reasonable.

The query on 1101 explicitly asks for `10.0.0.2`. I see that we're getting back the same record, but only _because we asked for the same ID_. I would have expected the query to fail, since we're asking for an existing record by ID (presumably using an ID allocated during a saga), but get back a record with a different IP than we asked for. Under what circumstances would we want that? It seems like we'd want to fail the query, because some higher-level software repeated a query, but with a different address. That seems indicative of a bug, but I might be missing something.

I guess it's possible for the software making this query to check that we got the address we actually asked for, but that also seems weird. If there's an explicit address being requested, I'd expect to either get back that address or to fail, not to succeed but with a different address.

Reply:
I agree with @bnaecker that this interface does not follow the principle of least surprise. Semantically, I'd expect failure or return of the address asked for instead.

Reply:
Nice, that looks like both a lot of work, and be very helpful! Thanks for taking that on.